### ISSUE TITLE
Fn is unusable (again) with kmonad

### ISSUE THREAD

**Issue by ZerdoX-x** [kmonad/kmonad#933]
Hi. I am on linux, using macbook air keyboard.

<details>
<summary>My config.kbd</summary>

```
(defcfg
	input	(device-file "/dev/input/by-path/platform-23510c000.spi-cs-0-event-kbd")
	output	(uinput-sink "KMonad kbd")
	fallthrough true
)

(defsrc
	esc	f1	f2	f3	f4	f5	f6	f7	f8	f9	f10	f11	f12
	grv	1	2	3	4	5	6	7	8	9	0	-	=	bspc
	tab	q	w	e	r	t	y	u	i	o	p	[	]	ret
	caps	a	s	d	f	g	h	j	k	l	;	'	\
	lsft	rctrl	z	x	c	v	b	n	m	,	.	/	rsft
	fn	lctl	lalt	lmet			spc			rmet	ralt
)

(defalias
	sxtr	(tap-next caps (layer-toggle extra))
	fn	(cmd-button "hyprctl dispatch exec \"tofi-drun | xargs hyprctl dispatch exec --\"")
)

(deflayer default
	esc	f1	f2	f3	f4	f5	f6	f7	f8	f9	f10	f11	f12
	grv	1	2	3	4	5	6	7	8	9	0	-	=	bspc
	tab	q	w	e	r	t	y	u	i	o	p	[	]	ret
	@sxtr	a	s	d	f	g	h	j	k	l	;	'	\
	lsft	rctrl	z	x	c	v	b	n	m	,	.	/	rsft
	fn	lctl	lalt	lmet			spc			rmet	ralt
)

(deflayer extra
	esc	f1	f2	f3	f4	f5	f6	f7	f8	f9	f10	f11	f12
	grv	1	2	3	4	5	6	7	8	9	0	-	=	bspc
	tab	1	2	3	4	5	6	7	esc	9	0	[	]	ret
	caps	ä	ä	_	_	_	_	left	down	up	right	'	\
	lsft	ä	!	@	#	$	%	^	&	\(	\)	/	rsft
	fn	lctl	lalt	lmet			spc			rmet	ralt
)


```

</details>

Wev with kmonad off:
```
[14:     wl_keyboard] key: serial: 47259; time: 7175272; key: 472; state: 1 (pressed)
                      sym: XF86Fn       (268964304), utf8: ''
[14:     wl_keyboard] key: serial: 47260; time: 7175393; key: 472; state: 0 (released)
                      sym: XF86Fn       (268964304), utf8: ''
```

But with kmonad running, fn button is not emitted.

Additional info:
I was able to use kmonad and bind something to fn using patch I made from https://github.com/kmonad/kmonad/pull/475/files. I did not upgrade quite a while and now I see kmonad is finally getting updates. So what is the correct way to handle this? And yes, I know about fn question in faq.

Also you can see `fn` alias for my launcher, it does not work either, though, now I am using hyprland key bindings, but i'll leave it as an option here too.

**Comment by jokesper** [comment]
Duplicate of #912.

**Comment by ZerdoX-x** [comment]
@jokesper how is this related?

**Comment by ZerdoX-x** [comment]
could you please re read my original post? kmonad just SWALLOWS fn.

**Comment by ZerdoX-x** [comment]
ah i see it now. could you please tell me how can I add my single missing key to source files so kmonad could use it? i tried manually same as i did it previously but it did not seem to work now

**Comment by jokesper** [comment]
Isn't the problem that KMonads keycodes only reach up to 256. This would be the same issue.

Independently of that, couldn't you just leave out the fn part and enable fallthrough true?

Sorry for the original terse answer

**Comment by ZerdoX-x** [comment]
yes i tried to remove fn from all layouts or leave it in source and put _, none works for me afaik. i'll double check later but i think this behavior should be somehow changed maybe

**Comment by ZerdoX-x** [comment]
> Isn't the problem that KMonads keycodes only reach up to 256. This would be the same issue.

yes, i am sure. the patch was extending keycodes, but it does not work now

**Comment by jokesper** [comment]
> i'll double check later but i think this behavior should be somehow changed maybe

Hopefully this works.

I will try to implement the `&<code>` feature @ david-janssen  suggested. I should have a PR ready till Friday (probably).

**Comment by jokesper** [comment]
> I will try to implement the `&<code>` feature @ david-janssen suggested

I haven't figured out how to integrate this with our macos code base, since we have to have an extra page entry.

Though your issue should be fixed by #935.

**Comment by ZerdoX-x** [comment]
> Though your issue should be fixed by https://github.com/kmonad/kmonad/pull/935.

Sadly no. I tried same configs. Any ideas?

**Comment by jokesper** [comment]
Does KMonad recognize fn (via `-l debug`)?

If not, it may be that those keys are on a different device (For the operating system not the actual keyboard).
You could also query the events by using `evdev <path-to-device>`

**Comment by ZerdoX-x** [comment]
> Does KMonad recognize fn (via `-l debug`)?

```
Opening KeySink
Registering Uinput device
Opening KeySource
Initiating ioctl grab
Launching process: emitter_proc

--------------------------------------------------------------------------------
Received event: Press <missing0x1ca>
Running hooks
Emitting: Press <missing0x1ca>
Registering untimed hook: 1

--------------------------------------------------------------------------------
Received event: Release <missing0x1ca>
Running hooks
Emitting: Release <missing0x1ca>
```


> If not, it may be that those keys are on a different device (For the operating system not the actual keyboard).

I already used fn with kmonad as a normal button, on same machine, M1 air. If really needed I could find the patch I used that times and working commit, but it will require some digging...

---

I can see you added Missing0x1ca to the code, perhaps, I should refer to it differently in my config?

**Comment by jokesper** [comment]
Interessting, according to `input-event-codes.h` `fn` should be `1d0`.
Though for you something like:
```clojure
(defalias
  fn missing0x1ca
)
```
and then using `@fn` instead would probably work,
I don't know if alias work in the `defsrc` though

**Comment by jokesper** [comment]
The real problem is a bug in #935.

**Comment by ZerdoX-x** [comment]
Works for me, cool, I'll remove my patch as soon as you merge it. Thanks! :snowflake:

**Comment by ZerdoX-x** [comment]
If anyone interested, here is my config for macbook air keyboard, mine one is [ISO variant](https://support.apple.com/en-us/102743)

<details>
<summary>config.kbd</summary>

```
(defcfg
	input	(device-file "/dev/input/by-path/platform-23510c000.spi-cs-0-event-kbd")
	output	(uinput-sink "KMonad kbd")
	fallthrough true
)

(defsrc
	esc	f1	f2	f3	f4	f5	f6	f7	f8	f9	f10	f11	f12
	grv	1	2	3	4	5	6	7	8	9	0	-	=	bspc
	tab	q	w	e	r	t	y	u	i	o	p	[	]	ret
	caps	a	s	d	f	g	h	j	k	l	;	'	\
	lsft	rctrl	z	x	c	v	b	n	m	,	.	/	rsft
	fn	lctl	lalt	lmet			spc			rmet	ralt
)

(defalias
	sxtr	(tap-next caps (layer-toggle extra))
)

(deflayer default
	esc	f1	f2	f3	f4	f5	f6	f7	f8	f9	f10	f11	f12
	grv	1	2	3	4	5	6	7	8	9	0	-	=	bspc
	tab	q	w	e	r	t	y	u	i	o	p	[	]	ret
	@sxtr	a	s	d	f	g	h	j	k	l	;	'	\
	lsft	rctrl	z	x	c	v	b	n	m	,	.	/	rsft
	fn	lctl	lalt	lmet			spc			rmet	ralt
)

(deflayer extra
	esc	f1	f2	f3	f4	f5	f6	f7	f8	f9	f10	f11	f12
	grv	1	2	3	4	5	6	7	8	9	0	-	=	bspc
	tab	1	2	3	4	5	6	7	esc	9	0	[	]	ret
	caps	ä	ä	_	_	_	_	left	down	up	right	'	\
	lsft	ä	!	@	#	$	%	^	&	\(	\)	/	rsft
	fn	lctl	lalt	lmet			spc			rmet	ralt
)
```

</details>

**Comment by ZerdoX-x** [comment]
Ooops.. It doesn't work, `-l debug` shows it's emitted as fn but wev doesn't show anything.
Though, I can now rebind it to rctrl and it emits it just fine, I just remap it to nonexistent key.

@jokesper want more details? I am satisfied with my workaround, but it is workaround.

**Comment by jokesper** [comment]
I think I know what might cause this.
We have to set capabilities to emit keycodes, it seems I forgot to increase them.

**Comment by ZerdoX-x** [comment]
Works now :)
